src/main/groovy/org/pih/warehouse/outbound/ImportPackingListItem.groovy

```diff
@@ -0,0 +1,98 @@
+        // If location is not found, but we provided bin location name, it means, the location was not found,
+        // and we want to indicate that, instead of falling back to default (null)
+        // without this, we would then search e.g. for quantity available to promise for a product in the default bin location
+        if (!location && source['binLocation'] && !source['binLocation'].equalsIgnoreCase(Constants.DEFAULT_BIN_LOCATION_NAME)) {
+            // We want to indicate, that a bin location for given name has not been found.
+            obj.binLocationFound = false
+        }
+        return location
+     })
+    Location binLocation
+
+    Integer quantityPicked
+
+    @BindUsing({ obj, source -> Person.findPersonByNameOrEmail(source['recipient'])})
+    Person recipient
+
+    /**
+     * Flag to indicate whether binLocation has been found and bound.
+     * We can't rely on binLocation = null, because we accept the null as DEFAULT,
+     * so we want to know if the binLocation is null, because we have not provided (or provided DEFAULT) the bin location name
+     * or the binLocation for such name does not exist.
+     * This can't be done in the BindUsing and rejectValue, because the errors would be cleared before running the validator logic
+     * The flag is set to false if a bin location is not found for given name in the BindUsing of binLocation
+     */
+    boolean binLocationFound = true
+
+
+    static constraints = {
+        rowId(blank: false)
+        palletName(nullable: true)
+        boxName(nullable: true, validator: { String boxName, ImportPackingListItem item ->
+            if (boxName && !item.palletName) {
+                return ['packLevel1.required']
+            }
+            return true
+        })
+        lotNumber(nullable: true, blank: true, validator: { String lotNumber, ImportPackingListItem item ->
+            InventoryItem inventoryItem = item.product?.getInventoryItem(lotNumber)
+            if (inventoryItem) {
+                return true
+            }
+            return ['inventoryItemNotFound', lotNumber, item.product?.productCode]
+        })
+        expirationDate(nullable: true)
+        binLocation(nullable: true)
+        quantityPicked(min: 0, validator: { Integer quantityPicked, ImportPackingListItem item ->
+            if (!item.binLocationFound) {
+                return ['binLocationNotFound']
+            }
+            ProductAvailabilityService productAvailabilityService = Holders.grailsApplication.mainContext.getBean(ProductAvailabilityService)
+            InventoryItem inventoryItem = item.product?.getInventoryItem(item.lotNumber)
+            if (!inventoryItem) {
+                return ['inventoryItemNotFound']
+            }
+            Location currentLocation = AuthService.currentLocation
+            Integer quantity = productAvailabilityService.getQuantityAvailableToPromiseForProductInBin(inventoryItem, currentLocation, item.binLocation)
```

Comment:
I understand the desire to have this here in teh constraints. But this seems dangerous. I think I'd prefer if the validateData was doing this validation. One benefit would be that we'd inject the productAvailabilityService into the Importer once vs once each time domain.validate() is called (likely more than once per domain). 

With that said, there an argument for allowing the domain to call the service. 
https://stackoverflow.com/a/3654488/136597
https://stackoverflow.com/a/2460029/136597

But perhaps we could inject the service into the domain once
https://stackoverflow.com/a/9699609/136597

Reply:
We decided not to perform the validation at import file level, hence we can't do it in the `validateData` method, as it would require us to refactor the whole thing now.
The `productAvailabilityService` is not injected per each item, since for each item we have the same memory reference to the service, so for each item it uses "the same" instance of product availability service.